protocols/kad/src/addresses.rs

```diff
@@ -22,67 +22,77 @@ use libp2p_core::Multiaddr;
 use smallvec::SmallVec;
 use std::fmt;
 
-/// List of addresses of a peer.
+/// A non-empty list of (unique) addresses of a peer in the routing table.
 #[derive(Clone)]
 pub struct Addresses {
     addrs: SmallVec<[Multiaddr; 6]>,
 }
 
 impl Addresses {
     /// Creates a new list of addresses.
-    pub fn new() -> Addresses {
-        Addresses {
-            addrs: SmallVec::new(),
-        }
+    pub fn new(addr: Multiaddr) -> Addresses {
+        let mut addrs = SmallVec::new();
+        addrs.push(addr);
+        Addresses { addrs }
+    }
+
+    /// Gets a reference to the first address in the list.
+    pub fn first(&self) -> &Multiaddr {
+        &self.addrs[0]
     }
 
-    /// Returns an iterator over the list of addresses.
+    /// Returns an iterator over the addresses.
     pub fn iter(&self) -> impl Iterator<Item = &Multiaddr> {
         self.addrs.iter()
     }
 
+    /// Returns the number of addresses in the list.
+    pub fn len(&self) -> usize {
+        self.addrs.len()
+    }
+
     /// Converts the addresses into a `Vec`.
     pub fn into_vec(self) -> Vec<Multiaddr> {
         self.addrs.into_vec()
     }
 
-    /// Returns true if the list of addresses is empty.
-    pub fn is_empty(&self) -> bool {
-        self.addrs.is_empty()
-    }
+    /// Removes the given address from the list.
+    ///
+    /// Returns true if the address was found and removed, false otherwise.
+    /// The last remaining address in the list cannot be remvoved.
+    ///
+    /// An address should only be removed if is determined to be invalid or
+    /// otherwise unreachable.
+    pub fn remove(&mut self, addr: &Multiaddr) -> bool {
+        if self.addrs.len() == 1 {
+            return false
```

Comment:
I am not very familiar with the protocol, but AFAICS an address is removed only if the peer can not be reached using it. So instead of keeping a peer with a defunct address, should the peer not be removed? Alternatively maybe adding a comment to `Addresses` which motivates this behaviour would be appreciated.

Reply:
I generally feel uneasy about removing addresses from peers (in `Kademlia::inject_addr_reach_failure`) because from my understanding many such errors may be transient.  There is a tension, of course. It is also desirable to remove unreachable addresses because they slow down queries, so the problem is when to decide when an address is considered "permanently" unreachable.

I did think about removing the peer from the routing table when the last address is deemed "unreachable", but removal of peers seems even more problematic, because it would mean that temporary loss of connectivity, whether benign or malicious, would "flush" a node's routing table, subverting the important property of the Kademlia protocol of hanging on to existing peers, at least until other peers demonstrate better / longer connectivity, causing least-recently connected nodes to be evicted. Thus I opted for a compromise for now where the last address of a peer is just not removed and hence a peer not removed from the routing table as a result if dialing errors (i.e. just like before with the only difference that I don't allow peers without any addresses in the routing table, hence hanging on to the last address).

Let me know if you have further thoughts / advice on the matter. If my reasoning and choice (for now) sounds good, I can improve the code commentary as suggested.

Reply:
Yeah,  it is better to only remove peers if a bucket is full and they do not answer a ping.

Reply:
I don't know if the code takes that into account, but keep in mind that when we receive an incoming, we don't "know" the address of the remote (it is determined through identify), yet we should maybe add it to the buckets already.